File: lwjgl/lesson26/Lesson26/Lesson26.java
/*
 *      This Code Was Created By Jeff Molofee and GB Schmick 2000
 *      A HUGE Thanks To Fredric Echols For Cleaning Up
 *      And Optimizing The Base Code, Making It More Flexible!
 *      If You've Found This Code Useful, Please Let Me Know.
 *      Visit Our Sites At www.tiptup.com and nehe.gamedev.net
 */

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;

import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.glu.GLU;
import org.lwjgl.opengl.glu.Sphere;
import org.lwjgl.devil.IL;
import org.lwjgl.input.Keyboard;

/**
 * @author Mark Bernard
 * date:    26-Jun-2004
 *
 * Port of NeHe's Lesson 26 to LWJGL
 * Title: Clipping & Reflections Using The Stencil Buffer
 * Uses version 0.9alpha of LWJGL http://www.lwjgl.org/
 *
 * Be sure that the LWJGL libraries are in your classpath
 *
 * Ported directly from the C++ version
 *
 * 2004-10-08: Updated to version 0.92alpha of LWJGL.
 * 2004-12-19: Updated to version 0.94alpha of LWJGL and to use
 *             DevIL for image loading.
 */
public class Lesson26 {
    private boolean done = false;
    private boolean fullscreen = false;
    private final String windowTitle = "NeHe's OpenGL Lesson 26 for LWJGL (Clipping & Reflections Using The Stencil Buffer)";
    private boolean f1 = false; // F1 key pressed
    private DisplayMode displayMode;

    //  Light Parameters
    static float lightAmbient[] = {0.7f, 0.7f, 0.7f, 1.0f};  // Ambient Light
    static float lightDiffuse[] = {1.0f, 1.0f, 1.0f, 1.0f};  // Diffuse Light
    static float lightPosition[] = {4.0f, 4.0f, 6.0f, 1.0f};  // Light Position
    ByteBuffer lightBuffer;
    ByteBuffer reflectBuffer;

    Sphere q;

    float xrot = 0.0f;                        // X Rotation
    float yrot = 0.0f;                        // Y Rotation
    float xrotspeed = 0.0f;                        // X Rotation Speed
    float yrotspeed = 0.0f;                        // Y Rotation Speed
    float zoom = -7.0f;                        // Depth Into The Screen
    float height = 2.0f;                        // Height Of Ball From Floor

    int texture[];                                 // 3 Textures

    public static void main(String args[]) {
        boolean fullscreen = false;
        if(args.length>0) {
            if(args[0].equalsIgnoreCase("fullscreen")) {
                fullscreen = true;
            }
        }

        Lesson26 l26 = new Lesson26();
        l26.run(fullscreen);
    }
    public void run(boolean fullscreen) {
        this.fullscreen = fullscreen;
        try {
            init();

            while (!done) {
                render();
                mainloop();
                Display.update();
            }
            cleanup();
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }
    private void mainloop() {
        if(Keyboard.isKeyDown(Keyboard.KEY_ESCAPE)) {       // Exit if Escape is pressed
            done = true;
        }
        if(Display.isCloseRequested()) {                     // Exit if window is closed
            done = true;
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_F1) && !f1) {    // Is F1 Being Pressed?
            f1 = true;                                      // Tell Program F1 Is Being Held
            switchMode();                                   // Toggle Fullscreen / Windowed Mode
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_F1)) {          // Is F1 Being Pressed?
            f1 = false;
        }

        if(Keyboard.isKeyDown(Keyboard.KEY_RIGHT))     yrotspeed += 0.08f;         // Right Arrow Pressed (Increase yrotspeed)
        if(Keyboard.isKeyDown(Keyboard.KEY_LEFT))      yrotspeed -= 0.08f;         // Left Arrow Pressed (Decrease yrotspeed)
        if(Keyboard.isKeyDown(Keyboard.KEY_DOWN))      xrotspeed += 0.08f;         // Down Arrow Pressed (Increase xrotspeed)
        if(Keyboard.isKeyDown(Keyboard.KEY_UP))        xrotspeed -= 0.08f;         // Up Arrow Pressed (Decrease xrotspeed)

        if(Keyboard.isKeyDown(Keyboard.KEY_A))          zoom +=0.05f;               // 'A' Key Pressed ... Zoom In
        if(Keyboard.isKeyDown(Keyboard.KEY_Z))          zoom -=0.05f;               // 'Z' Key Pressed ... Zoom Out

        if(Keyboard.isKeyDown(Keyboard.KEY_PRIOR))     height +=0.03f;             // Page Up Key Pressed Move Ball Up
        if(Keyboard.isKeyDown(Keyboard.KEY_NEXT))      height -=0.03f;             // Page Down Key Pressed Move Ball Down
    }

    private void switchMode() {
        fullscreen = !fullscreen;
        try {
            Display.setFullscreen(fullscreen);
        }
        catch(Exception e) {
            e.printStackTrace();
        }
    }

    private void createWindow() throws Exception {
        Display.setFullscreen(fullscreen);
        DisplayMode d[] = Display.getAvailableDisplayModes();
        for (int i = 0; i < d.length; i++) {
            if (d[i].getWidth() == 640
                && d[i].getHeight() == 480
                && d[i].getBitsPerPixel() == 32) {
                displayMode = d[i];
                break;
            }
        }
        Display.setDisplayMode(displayMode);
        Display.setTitle(windowTitle);
        Display.create();
    }
    private void init() throws Exception {
        createWindow();
        IL.create();

        initGL();

        loadTextures();
    }

    private void initGL() {                                         // All Setup For OpenGL Goes Here
        int width = 640;
        int height = 480;
        GL11.glShadeModel(GL11.GL_SMOOTH);                            // Enable Smooth Shading
        GL11.glClearColor(0.2f, 0.5f, 1.0f, 1.0f);               // Background
        GL11.glClearDepth(1.0f);                                 // Depth Buffer Setup
        GL11.glClearStencil(0);                                  // Clear The Stencil Buffer To 0
        GL11.glEnable(GL11.GL_DEPTH_TEST);                            // Enables Depth Testing
        GL11.glDepthFunc(GL11.GL_LEQUAL);                             // The Type Of Depth Testing To Do
        GL11.glHint(GL11.GL_PERSPECTIVE_CORRECTION_HINT, GL11.GL_NICEST);  // Really Nice Perspective Calculations
        GL11.glEnable(GL11.GL_TEXTURE_2D);                            // Enable 2D Texture Mapping

        reflectBuffer = ByteBuffer.allocateDirect(32);
        reflectBuffer.order(ByteOrder.nativeOrder());
        lightBuffer = ByteBuffer.allocateDirect(16);
        lightBuffer.order(ByteOrder.nativeOrder());
        GL11.glLight(GL11.GL_LIGHT0, GL11.GL_AMBIENT, (FloatBuffer)lightBuffer.asFloatBuffer().put(lightAmbient).flip());              // Setup The Ambient Light
        GL11.glLight(GL11.GL_LIGHT0, GL11.GL_DIFFUSE, (FloatBuffer)lightBuffer.asFloatBuffer().put(lightDiffuse).flip());              // Setup The Diffuse Light
        GL11.glLight(GL11.GL_LIGHT0, GL11.GL_POSITION, (FloatBuffer)lightBuffer.asFloatBuffer().put(lightPosition).flip());         // Position The Light

        GL11.glEnable(GL11.GL_LIGHT0);                                // Enable Light 0
        GL11.glEnable(GL11.GL_LIGHTING);                              // Enable Lighting

        q = new Sphere();
        q.setNormals(GL11.GL_SMOOTH);
        q.setTextureFlag(true);

        GL11.glTexGeni(GL11.GL_S, GL11.GL_TEXTURE_GEN_MODE, GL11.GL_SPHERE_MAP);    // Set Up Sphere Mapping
        GL11.glTexGeni(GL11.GL_T, GL11.GL_TEXTURE_GEN_MODE, GL11.GL_SPHERE_MAP);    // Set Up Sphere Mapping

        GL11.glViewport(0,0,width,height);                       // Reset The Current Viewport

        GL11.glMatrixMode(GL11.GL_PROJECTION);                        // Select The Projection Matrix
        GL11.glLoadIdentity();                                   // Reset The Projection Matrix

        // Calculate The Aspect Ratio Of The Window
        GLU.gluPerspective(45.0f,
                (float) displayMode.getWidth() / (float) displayMode.getHeight(),
                0.1f,100.0f);

        GL11.glMatrixMode(GL11.GL_MODELVIEW);                         // Select The Modelview Matrix
        GL11.glLoadIdentity();                                   // Reset The Modelview Matrix
    }

    private void drawObject() {                                      // Draw Our Ball
        GL11.glColor3f(1.0f, 1.0f, 1.0f);                        // Set Color To White
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[1]);           // Select Texture 2 (1)
        q.draw(0.35f, 32, 16);                        // Draw First Sphere

        GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[2]);           // Select Texture 3 (2)
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 0.4f);                  // Set Color To White With 40% Alpha
        GL11.glEnable(GL11.GL_BLEND);                                 // Enable Blending
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE);                  // Set Blending Mode To Mix Based On SRC Alpha
        GL11.glEnable(GL11.GL_TEXTURE_GEN_S);                         // Enable Sphere Mapping
        GL11.glEnable(GL11.GL_TEXTURE_GEN_T);                         // Enable Sphere Mapping

        q.draw(0.35f, 32, 16);                              // Draw Another Sphere Using New Texture
                                                            // Textures Will Mix Creating A MultiTexture Effect (Reflection)
        GL11.glDisable(GL11.GL_TEXTURE_GEN_S);                        // Disable Sphere Mapping
        GL11.glDisable(GL11.GL_TEXTURE_GEN_T);                        // Disable Sphere Mapping
        GL11.glDisable(GL11.GL_BLEND);                                // Disable Blending
    }

    private void drawFloor() {                                        // Draws The Floor
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, texture[0]);           // Select Texture 1 (0)
        GL11.glBegin(GL11.GL_QUADS);                                  // Begin Drawing A Quad
            GL11.glNormal3f(0.0f, 1.0f, 0.0f);                      // Normal Pointing Up
            GL11.glTexCoord2f(0.0f, 1.0f);                   // Bottom Left Of Texture
            GL11.glVertex3f(-2.0f, 0.0f, 2.0f);                 // Bottom Left Corner Of Floor

            GL11.glTexCoord2f(0.0f, 0.0f);                   // Top Left Of Texture
            GL11.glVertex3f(-2.0f, 0.0f,-2.0f);                 // Top Left Corner Of Floor

            GL11.glTexCoord2f(1.0f, 0.0f);                   // Top Right Of Texture
            GL11.glVertex3f( 2.0f, 0.0f,-2.0f);                 // Top Right Corner Of Floor

            GL11.glTexCoord2f(1.0f, 1.0f);                   // Bottom Right Of Texture
            GL11.glVertex3f( 2.0f, 0.0f, 2.0f);                 // Bottom Right Corner Of Floor
        GL11.glEnd();                                            // Done Drawing The Quad
    }

    private void render() {
        // Clear Screen, Depth Buffer & Stencil Buffer
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT | GL11.GL_STENCIL_BUFFER_BIT);

        // Clip Plane Equations
        double eqr[] = {0.0f,-1.0f, 0.0f, 0.0f};            // Plane Equation To Use For The Reflected Objects

        GL11.glLoadIdentity();                                   // Reset The Modelview Matrix
        GL11.glTranslatef(0.0f, -0.6f, zoom);                    // Zoom And Raise Camera Above The Floor (Up 0.6 Units)
        GL11.glColorMask(false, false, false, false);                               // Set Color Mask
        GL11.glEnable(GL11.GL_STENCIL_TEST);                          // Enable Stencil Buffer For "marking" The Floor
        GL11.glStencilFunc(GL11.GL_ALWAYS, 1, 255);                     // Always Passes, 1 Bit Plane, 1 As Mask
        GL11.glStencilOp(GL11.GL_KEEP, GL11.GL_KEEP, GL11.GL_REPLACE);          // We Set The Stencil Buffer To 1 Where We Draw Any Polygon
                                                            // Keep If Test Fails, Keep If Test Passes But Buffer Test Fails
                                                            // Replace If Test Passes
        GL11.glDisable(GL11.GL_DEPTH_TEST);                           // Disable Depth Testing
        drawFloor();                                        // Draw The Floor (Draws To The Stencil Buffer)
                                                            // We Only Want To Mark It In The Stencil Buffer
        GL11.glEnable(GL11.GL_DEPTH_TEST);                            // Enable Depth Testing
        GL11.glColorMask(true, true, true, true);                               // Set Color Mask to TRUE, TRUE, TRUE, TRUE
        GL11.glStencilFunc(GL11.GL_EQUAL, 1, 255);                      // We Draw Only Where The Stencil Is 1
                                                            // (I.E. Where The Floor Was Drawn)
        GL11.glStencilOp(GL11.GL_KEEP, GL11.GL_KEEP, GL11.GL_KEEP);             // Don't Change The Stencil Buffer
        GL11.glEnable(GL11.GL_CLIP_PLANE0);                           // Enable Clip Plane For Removing Artifacts
                                                            // (When The Object Crosses The Floor)
        GL11.glClipPlane(GL11.GL_CLIP_PLANE0, (DoubleBuffer)reflectBuffer.asDoubleBuffer().put(eqr).flip());                   // Equation For Reflected Objects
        GL11.glPushMatrix();                                     // Push The Matrix Onto The Stack
            GL11.glScalef(1.0f, -1.0f, 1.0f);                    // Mirror Y Axis
            GL11.glLight(GL11.GL_LIGHT0, GL11.GL_POSITION, (FloatBuffer)lightBuffer.asFloatBuffer().put(lightPosition).flip());    // Set Up Light0
            GL11.glTranslatef(0.0f, height, 0.0f);               // Position The Object
            GL11.glRotatef(xrot, 1.0f, 0.0f, 0.0f);              // Rotate Local Coordinate System On X Axis
            GL11.glRotatef(yrot, 0.0f, 1.0f, 0.0f);              // Rotate Local Coordinate System On Y Axis
            drawObject();                                   // Draw The Sphere (Reflection)
        GL11.glPopMatrix();                                      // Pop The Matrix Off The Stack
        GL11.glDisable(GL11.GL_CLIP_PLANE0);                          // Disable Clip Plane For Drawing The Floor
        GL11.glDisable(GL11.GL_STENCIL_TEST);                         // We Don't Need The Stencil Buffer Any More (Disable)
        GL11.glLight(GL11.GL_LIGHT0, GL11.GL_POSITION, (FloatBuffer)lightBuffer.asFloatBuffer().put(lightPosition).flip());        // Set Up Light0 Position
        GL11.glEnable(GL11.GL_BLEND);                                 // Enable Blending (Otherwise The Reflected Object Wont Show)
        GL11.glDisable(GL11.GL_LIGHTING);                             // Since We Use Blending, We Disable Lighting
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 0.8f);                  // Set Color To White With 80% Alpha
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);  // Blending Based On Source Alpha And 1 Minus Dest Alpha
        drawFloor();                                        // Draw The Floor To The Screen
        GL11.glEnable(GL11.GL_LIGHTING);                              // Enable Lighting
        GL11.glDisable(GL11.GL_BLEND);                                // Disable Blending
        GL11.glTranslatef(0.0f, height, 0.0f);                   // Position The Ball At Proper Height
        GL11.glRotatef(xrot, 1.0f, 0.0f, 0.0f);                  // Rotate On The X Axis
        GL11.glRotatef(yrot, 0.0f, 1.0f, 0.0f);                  // Rotate On The Y Axis
        drawObject();                                       // Draw The Ball
        xrot += xrotspeed;                                  // Update X Rotation Angle By xrotspeed
        yrot += yrotspeed;                                  // Update Y Rotation Angle By yrotspeed
        GL11.glFlush();                                          // Flush The GL Pipeline
    }

    private void cleanup() {
        Display.destroy();
    }

    private void loadTextures() {                                   // Load Bitmaps And Convert To Textures
        texture = new int[3];
        texture[0] = loadTexture("Data/EnvWall.bmp");
        texture[1] = loadTexture("Data/Ball.bmp");
        texture[2] = loadTexture("Data/EnvRoll.bmp");
    }

    /**
     * Texture loading directly from LWJGL examples
     */
    private int loadTexture(String path) {
        IntBuffer image = ByteBuffer.allocateDirect(4).order(ByteOrder.nativeOrder()).asIntBuffer();
        IL.ilGenImages(1, image);
        IL.ilBindImage(image.get(0));
        IL.ilLoadImage(path);
        IL.ilConvertImage(IL.IL_RGB, IL.IL_BYTE);
        ByteBuffer scratch = ByteBuffer.allocateDirect(IL.ilGetInteger(IL.IL_IMAGE_WIDTH) * IL.ilGetInteger(IL.IL_IMAGE_HEIGHT) * 3);
        IL.ilCopyPixels(0, 0, 0, IL.ilGetInteger(IL.IL_IMAGE_WIDTH), IL.ilGetInteger(IL.IL_IMAGE_HEIGHT), 1, IL.IL_RGB, IL.IL_BYTE, scratch);
        
        // Create A IntBuffer For Image Address In Memory
        IntBuffer buf = ByteBuffer.allocateDirect(4).order(ByteOrder.nativeOrder()).asIntBuffer();
        GL11.glGenTextures(buf); // Create Texture In OpenGL

        GL11.glBindTexture(GL11.GL_TEXTURE_2D, buf.get(0));
        // Typical Texture Generation Using Data From The Image

        // Linear Filtering
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_LINEAR);
        // Linear Filtering
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR);
        // Generate The Texture
        GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL11.GL_RGB, IL.ilGetInteger(IL.IL_IMAGE_WIDTH), 
                IL.ilGetInteger(IL.IL_IMAGE_HEIGHT), 0, GL11.GL_RGB, GL11.GL_UNSIGNED_BYTE, scratch);

        return buf.get(0); // Return Image Address In Memory
    }
}

File: linux/lesson26/ztvC3/lesson26/Lesson26.cpp
/*		This code has been created by Banu Cosmin aka Choko - 20 may 2000
 *		and uses NeHe tutorials as a starting point (window initialization,
 *		texture loading, GL initialization and code for keypresses) - very good
 *		tutorials, Jeff. If anyone is interested about the presented algorithm
 *		please e-mail me at boct@romwest.ro
 *
 *		Code Commmenting And Clean Up By Jeff Molofee ( NeHe )
 *		NeHe Productions	...		http://nehe.gamedev.net
 */

#include <GL/gl.h>										// Header File For The OpenGL32 Library
#include <GL/glu.h>										// Header File For The GLu32 Library
#include <GL/glut.h>									// Header File For The Glaux Library
#include <stdio.h>
#include <stdlib.h>										// Header File For Standard Input / Output
#include "image.h"


// Light Parameters
static GLfloat LightAmb[] = {0.7f, 0.7f, 0.7f, 1.0f};	// Ambient Light
static GLfloat LightDif[] = {1.0f, 1.0f, 1.0f, 1.0f};	// Diffuse Light
static GLfloat LightPos[] = {4.0f, 4.0f, 6.0f, 1.0f};	// Light Position

GLUquadricObj	*q;										// Quadratic For Drawing A Sphere

GLfloat		xrot		=  0.0f;						// X Rotation
GLfloat		yrot		=  0.0f;						// Y Rotation
GLfloat		xrotspeed	=  0.0f;						// X Rotation Speed
GLfloat		yrotspeed	=  0.0f;						// Y Rotation Speed
GLfloat		zoom		= -7.0f;						// Depth Into The Screen
GLfloat		height		=  2.0f;						// Height Of Ball From Floor

GLuint		texture[3];									// 3 Textures

GLvoid Reshape(GLsizei width, GLsizei height)		// Resize And Initialize The GL Window
{
	if (height==0)										// Prevent A Divide By Zero By
	{
		height=1;										// Making Height Equal One
	}

	glViewport(0,0,width,height);						// Reset The Current Viewport

	glMatrixMode(GL_PROJECTION);						// Select The Projection Matrix
	glLoadIdentity();									// Reset The Projection Matrix

	// Calculate The Aspect Ratio Of The Window
	gluPerspective(45.0f,(GLfloat)width/(GLfloat)height,0.1f,100.0f);

	glMatrixMode(GL_MODELVIEW);							// Select The Modelview Matrix
	glLoadIdentity();									// Reset The Modelview Matrix
}

int LoadGLTextures()                                    // Load Bitmaps And Convert To Textures
{
    int Status=false;									// Status Indicator
    IMAGE *TextureImage[3];					// Create Storage Space For The Textures

    if ((TextureImage[0]=ImageLoad("Data/Envwall.rgb")) &&// Load The Floor Texture
        (TextureImage[1]=ImageLoad("Data/Ball.rgb")) &&	// Load the Light Texture
        (TextureImage[2]=ImageLoad("Data/Envroll.rgb")))	// Load the Wall Texture
	{
		Status=true;									// Set The Status To true
		glGenTextures(3, &texture[0]);					// Create The Texture
		for (int loop=0; loop<3; loop++)				// Loop Through 5 Textures
		{
			glBindTexture(GL_TEXTURE_2D, texture[loop]);
			glTexImage2D(GL_TEXTURE_2D, 0, 3, TextureImage[loop]->sizeX, TextureImage[loop]->sizeY, 0, GL_RGB, GL_UNSIGNED_BYTE, TextureImage[loop]->data);
			glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MIN_FILTER,GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MAG_FILTER,GL_LINEAR);
		}
		for (int loop=0; loop<3; loop++)					// Loop Through 5 Textures
		{
			if (TextureImage[loop])						// If Texture Exists
			{
				if (TextureImage[loop]->data)			// If Texture Image Exists
				{
					free(TextureImage[loop]->data);		// Free The Texture Image Memory
				}
				free(TextureImage[loop]);				// Free The Image Structure
			}
		}
	}
	return Status;										// Return The Status
}

int InitGL(GLvoid)										// All Setup For OpenGL Goes Here
{
	if (!LoadGLTextures())								// If Loading The Textures Failed
	{
		return false;									// Return false
	}
	glShadeModel(GL_SMOOTH);							// Enable Smooth Shading
	glClearColor(0.2f, 0.5f, 1.0f, 1.0f);				// Background
	glClearDepth(1.0f);									// Depth Buffer Setup
	glClearStencil(0);									// Clear The Stencil Buffer To 0
	glEnable(GL_DEPTH_TEST);							// Enables Depth Testing
	glDepthFunc(GL_LEQUAL);								// The Type Of Depth Testing To Do
	glHint(GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST);	// Really Nice Perspective Calculations
	glEnable(GL_TEXTURE_2D);							// Enable 2D Texture Mapping

	glLightfv(GL_LIGHT0, GL_AMBIENT, LightAmb);			// Set The Ambient Lighting For Light0
	glLightfv(GL_LIGHT0, GL_DIFFUSE, LightDif);			// Set The Diffuse Lighting For Light0
	glLightfv(GL_LIGHT0, GL_POSITION, LightPos);		// Set The Position For Light0

	glEnable(GL_LIGHT0);								// Enable Light 0
	glEnable(GL_LIGHTING);								// Enable Lighting

	q = gluNewQuadric();								// Create A New Quadratic
	gluQuadricNormals(q, GL_SMOOTH);					// Generate Smooth Normals For The Quad
	gluQuadricTexture(q, GL_TRUE);						// Enable Texture Coords For The Quad

	glTexGeni(GL_S, GL_TEXTURE_GEN_MODE, GL_SPHERE_MAP);	// Set Up Sphere Mapping
	glTexGeni(GL_T, GL_TEXTURE_GEN_MODE, GL_SPHERE_MAP);	// Set Up Sphere Mapping

	return true;										// Initialization Went OK
}

void DrawObject()										// Draw Our Ball
{
	glColor3f(1.0f, 1.0f, 1.0f);						// Set Color To White
	glBindTexture(GL_TEXTURE_2D, texture[1]);			// Select Texture 2 (1)
	gluSphere(q, 0.35f, 32, 16);						// Draw First Sphere

	glBindTexture(GL_TEXTURE_2D, texture[2]);			// Select Texture 3 (2)
	glColor4f(1.0f, 1.0f, 1.0f, 0.4f);					// Set Color To White With 40% Alpha
	glEnable(GL_BLEND);									// Enable Blending
	glBlendFunc(GL_SRC_ALPHA, GL_ONE);					// Set Blending Mode To Mix Based On SRC Alpha
	glEnable(GL_TEXTURE_GEN_S);							// Enable Sphere Mapping
	glEnable(GL_TEXTURE_GEN_T);							// Enable Sphere Mapping

	gluSphere(q, 0.35f, 32, 16);						// Draw Another Sphere Using New Texture
														// Textures Will Mix Creating A MultiTexture Effect (Reflection)
	glDisable(GL_TEXTURE_GEN_S);						// Disable Sphere Mapping
	glDisable(GL_TEXTURE_GEN_T);						// Disable Sphere Mapping
	glDisable(GL_BLEND);								// Disable Blending
}

void DrawFloor()										// Draws The Floor
{
	glBindTexture(GL_TEXTURE_2D, texture[0]);			// Select Texture 1 (0)
	glBegin(GL_QUADS);									// Begin Drawing A Quad
		glNormal3f(0.0, 1.0, 0.0);						// Normal Pointing Up
			glTexCoord2f(0.0f, 1.0f);					// Bottom Left Of Texture
			glVertex3f(-2.0, 0.0, 2.0);					// Bottom Left Corner Of Floor

			glTexCoord2f(0.0f, 0.0f);					// Top Left Of Texture
			glVertex3f(-2.0, 0.0,-2.0);					// Top Left Corner Of Floor

			glTexCoord2f(1.0f, 0.0f);					// Top Right Of Texture
			glVertex3f( 2.0, 0.0,-2.0);					// Top Right Corner Of Floor

			glTexCoord2f(1.0f, 1.0f);					// Bottom Right Of Texture
			glVertex3f( 2.0, 0.0, 2.0);					// Bottom Right Corner Of Floor
	glEnd();											// Done Drawing The Quad
}

static void Display(GLvoid)									// Draw Everything
{
	// Clear Screen, Depth Buffer & Stencil Buffer
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);

	// Clip Plane Equations
	double eqr[] = {0.0f,-1.0f, 0.0f, 0.0f};			// Plane Equation To Use For The Reflected Objects

	glLoadIdentity();									// Reset The Modelview Matrix
	glTranslatef(0.0f, -0.6f, zoom);					// Zoom And Raise Camera Above The Floor (Up 0.6 Units)
	glColorMask(0,0,0,0);								// Set Color Mask
	glEnable(GL_STENCIL_TEST);							// Enable Stencil Buffer For "marking" The Floor
	glStencilFunc(GL_ALWAYS, 1, 1);						// Always Passes, 1 Bit Plane, 1 As Mask
	glStencilOp(GL_KEEP, GL_KEEP, GL_REPLACE);			// We Set The Stencil Buffer To 1 Where We Draw Any Polygon
														// Keep If Test Fails, Keep If Test Passes But Buffer Test Fails
														// Replace If Test Passes
	glDisable(GL_DEPTH_TEST);							// Disable Depth Testing
	DrawFloor();										// Draw The Floor (Draws To The Stencil Buffer)
														// We Only Want To Mark It In The Stencil Buffer
	glEnable(GL_DEPTH_TEST);							// Enable Depth Testing
	glColorMask(1,1,1,1);								// Set Color Mask to true, true, true, true
	glStencilFunc(GL_EQUAL, 1, 1);						// We Draw Only Where The Stencil Is 1
														// (I.E. Where The Floor Was Drawn)
	glStencilOp(GL_KEEP, GL_KEEP, GL_KEEP);				// Don't Change The Stencil Buffer
	glEnable(GL_CLIP_PLANE0);							// Enable Clip Plane For Removing Artifacts
														// (When The Object Crosses The Floor)
	glClipPlane(GL_CLIP_PLANE0, eqr);					// Equation For Reflected Objects
	glPushMatrix();										// Push The Matrix Onto The Stack
		glScalef(1.0f, -1.0f, 1.0f);					// Mirror Y Axis
		glLightfv(GL_LIGHT0, GL_POSITION, LightPos);	// Set Up Light0
		glTranslatef(0.0f, height, 0.0f);				// Position The Object
		glRotatef(xrot, 1.0f, 0.0f, 0.0f);				// Rotate Local Coordinate System On X Axis
		glRotatef(yrot, 0.0f, 1.0f, 0.0f);				// Rotate Local Coordinate System On Y Axis
		DrawObject();									// Draw The Sphere (Reflection)
	glPopMatrix();										// Pop The Matrix Off The Stack
	glDisable(GL_CLIP_PLANE0);							// Disable Clip Plane For Drawing The Floor
	glDisable(GL_STENCIL_TEST);							// We Don't Need The Stencil Buffer Any More (Disable)
	glLightfv(GL_LIGHT0, GL_POSITION, LightPos);		// Set Up Light0 Position
	glEnable(GL_BLEND);									// Enable Blending (Otherwise The Reflected Object Wont Show)
	glDisable(GL_LIGHTING);								// Since We Use Blending, We Disable Lighting
	glColor4f(1.0f, 1.0f, 1.0f, 0.8f);					// Set Color To White With 80% Alpha
	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);	// Blending Based On Source Alpha And 1 Minus Dest Alpha
	DrawFloor();										// Draw The Floor To The Screen
	glEnable(GL_LIGHTING);								// Enable Lighting
	glDisable(GL_BLEND);								// Disable Blending
	glTranslatef(0.0f, height, 0.0f);					// Position The Ball At Proper Height
	glRotatef(xrot, 1.0f, 0.0f, 0.0f);					// Rotate On The X Axis
	glRotatef(yrot, 0.0f, 1.0f, 0.0f);					// Rotate On The Y Axis
	DrawObject();										// Draw The Ball
	xrot += xrotspeed;									// Update X Rotation Angle By xrotspeed
	yrot += yrotspeed;									// Update Y Rotation Angle By yrotspeed
	glFlush();											// Flush The GL Pipeline

	glutSwapBuffers();
}

static void Key( unsigned char key, int x, int y )
{
   switch (key) {
      case 27:
         exit(0);
         break;

      case 'A':
		zoom +=0.05;
      	break;

      case 'Z':
		zoom -=0.05f;
      	break;

   }
   glutPostRedisplay();
}


static void SpecialKey( int key, int x, int y )
{
   switch (key) {
      case GLUT_KEY_UP:
         xrotspeed += 0.08f;
         break;
      case GLUT_KEY_DOWN:
         xrotspeed -= 0.08f;
         break;
      case GLUT_KEY_LEFT:
         yrotspeed -= 0.08f;
         break;
      case GLUT_KEY_RIGHT:
         yrotspeed += 0.08f;
         break;
      case GLUT_KEY_PAGE_UP:
         height +=0.03f;
         break;
      case GLUT_KEY_PAGE_DOWN:
         height -=0.03f;
         break;
   }
   glutPostRedisplay();
}




int main( int argc, char *argv[] )
{
   glutInit( &argc, argv );
   glutInitWindowPosition( 0, 0 );
   glutInitWindowSize( 640, 480 );
   glutInitDisplayMode( GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH | GLUT_STENCIL );
   glutCreateWindow(argv[0]);
   glutReshapeFunc( Reshape );
   glutKeyboardFunc( Key );
   glutSpecialFunc( SpecialKey );
   glutDisplayFunc( Display );
   InitGL();
   glutMainLoop();
   return 0;
}
